src/librustc/mir/interpret/error.rs

```diff
@@ -310,20 +313,28 @@ impl<O: fmt::Debug> fmt::Debug for PanicMessage<O> {
 }
 
 #[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
-pub enum InterpError<'tcx> {
-    /// This variant is used by machines to signal their own errors that do not
-    /// match an existing variant.
-    MachineError(String),
+pub enum InvalidProgramInfo<'tcx> {
+    /// Resolution can fail if we are in a too generic context
+    TooGeneric,
+    /// Cannot compute this constant because it depends on another one
+    /// which already produced an error
+    ReferencedConstant,
+    /// Abort in case type errors are reached
+    TypeckError,
+    Layout(layout::LayoutError<'tcx>),
+}
 
-    /// Not actually an interpreter error -- used to signal that execution has exited
-    /// with the given status code.  Used by Miri, but not by CTFE.
-    Exit(i32),
+#[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
+pub enum UndefinedBehaviourInfo {
+    Unreachable,
+}
 
+#[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
+pub enum UnsupportedInfo<'tcx> {
```

Comment:
@RalfJung I have not implemented anything. But I have moved some enum variants into the new enums. Pretty sure most of them are wrong though.

Reply:
Please decide for one of these two strategies:

* leave all "unclassified" error variants directly in `InterpError`, and move things to the `*Info` enums once we agreed where they should go.
* or move all variants to `UnsupportedInfo` and we can then triage them from there.

Currently you seem to have done a mixture of that. Also please have a comment like `// -- the following variants have not been classified yet --`.

Reply:
The second strategy is what I have done now. All the remaining variants are now in `UnsupportedInfo `

Reply:
@RalfJung Can we discuss where the variants should move to in this PR, or in the actual issue?

Reply:
I think it is best to do this in batches -- as in, for this first PR, I think the classification you have now is actually a good start, `InvalidProgramInfo` and `ResourceExhaustionInfo` look complete and `UndefinedBehaviourInfo` has one representative example. So, it might make sense as the next step to try and get everything to compile again. What do you think?

One remark for this: Most of the error-causing places use the [`err!` macro](https://github.com/rust-lang/rust/blob/4268e7ee22935f086b856ef0063a9e22b49aeddb/src/librustc/mir/interpret/mod.rs#L4). Since you moved most variants to `Unsupported`, it probably makes sense to adjust that macro to use `Unsupported`. That should make porting all the code much easier.

Reply:
> One remark for this: Most of the error-causing places use the err! macro. Since you moved most variants to Unsupported, it probably makes sense to adjust that macro to use Unsupported. That should make porting all the code much easier.

I saw this after making the changes :(. Do I change `err!` now? Or do I do it in the next PR?

Reply:
Yes please change it. It would be better to keep this noise out of the diff.

Reply:
And also `err!` is also used for the other variants. So we would have to create multiple versions of `err!` right?

Reply:
Yes, good idea.

I think in the end we want `err_ub!` (for `UndefinedBehavior`), `err_unsup!` (for `UnsupportedOp`), and maybe more (the other variants are not thrown a lot I think)?

Well, `err_unsup` is an awful name, but I cannot come up with anything better for now. But for this PR, you can leave `err!` to mean `UnsupportedOp`, and just add `err_ub!` for the one variant that we did port (`Unreachable`).

Reply:
We also need a `err_inval` for `InvalidProgram`.

Reply:
I am confused about changing the `err!` macro to use `Unsupported`.

From this:
```rust
macro_rules! err {
    ($($tt:tt)*) => { Err($crate::mir::interpret::InterpError::$($tt)*.into()) };
}
```

I have changed the macro to this
```rust
macro_rules! err {
    ($($tt:tt)*) => { Err($crate::mir::interpret::InterpError::Unsupported::$($tt)*.into()) };
}
```

This the error I get:
```
error[E0433]: failed to resolve: `Unsupported` is a variant, not a module
   --> src/librustc/mir/interpret/mod.rs:5:64
    |
4   | / macro_rules! err {
5   | |     ($($tt:tt)*) => { Err($crate::mir::interpret::InterpError::Unsupported::$($tt)*.into()) };
    | |                                                                ^^^^^^^^^^^ `Unsupported` is a variant, not a module
6   | | }
    | |_- in this expansion of `err!`
    | 
   ::: src/librustc/mir/interpret/pointer.rs:202:13
    |
202 |               err!(PointerOutOfBounds { ptr: self.erase_tag(),msg,allocation_size })
    |               ---------------------------------------------------------------------- in this macro invocation

^C  Building [===============================================>       ] 137/155: rustc   
```

How do I fix this?

Reply:
> We also need a err_inval for InvalidProgram.

IIRC we don't trigger this error very often, but sure, seems reasonable for consistency.

> How do I fix this?

The macro should be more like
```rust
Err($crate::mir::interpret::InterpError::Unsupported($crate::mir::interpret::InterpError::UnsupportedInfo::$($tt)*).into())
```

Your macro expands to things like `InterpError::Unsupported::PointerOutOfBounds`, which does not make sense.

Reply:
Ah, stupid of me. Thanks!